Introduce support for packit builds

mwperina opened the pull request:
Introduces ability to build RPMs using packit and run integration tests within testing farm.

Current Github CI to build RPMs and run integrations tests will be remove in follow up PR.

Fixes: https://github.com/containers/bluechi/issues/452

mwperina committed:
Introduce support for packit builds

- Adapt our RPM build scripts to be usable in packit
- Add support to build RPMs for the most important architectures when PR
  is added/updated
- Add support to build RPMs for all supported platforms and save produced
  RPMs into our COPR repository

Fixes: https://github.com/containers/bluechi/issues/452
Signed-off-by: Martin Perina <mperina@redhat.com>

mwperina committed:
Introduce support for packit testing farm

- Adapt our integration tests to be executable by packit in the testing
  farm
- Add support to execute integration tests after RPMs are built when PR
  is added/updated

Fixes: https://github.com/containers/bluechi/issues/452
Signed-off-by: Martin Perina <mperina@redhat.com>

Yarboa commented on tests/conftest.py:
```diff
@@ -1,5 +1,17 @@
+def pytest_collection_modifyitems(items):
+    for item in items:
+        if item.get_closest_marker('timeout') is None:
+            item.add_marker(pytest.mark.timeout(30))
```
Please consider change it to 45
Please refer the discussions here 
https://github.com/containers/bluechi/pull/516

engelmi commented on tests/containers/integration-test-packit:
```diff
@@ -0,0 +1,24 @@
+FROM quay.io/bluechi/integration-test-base:latest
```
Couldn't we reuse the `integration-test-snapshot` image and pass in the `$copr_repo` also there? (which would be`@centos-automotive-sig/bluechi-snapshot`)

mwperina replied:
For each PR packit creates new temporary COPR repo for unmerged RPMs. For this PR it's https://copr.fedorainfracloud.org/coprs/packit/eclipse-bluechi-bluechi-488/ that's why I need to pass different copr repo for each PR.

Only after merge RPMs will be pushed to `@centos-automotive-sig/bluechi-snapshot`

mwperina replied:
Fixed

engelmi commented on tests/scripts/tests-setup.sh:
```diff
@@ -3,12 +3,32 @@
-   export ARCH="--arch=$(uname -m)"
+# Install required dependencies only when asked
+if [ "$INSTALL_DEPS" == "yes" ]; then
+    dnf install \
```
Isn't it possible to define those dependencies in the `.packit.yml`? E.g. for `srpm_build_deps`.

Looking at the GH CI, `podman` (python bindings), `pytest` and `pytest-timeout` are installed via pip - I assume the versions installed via `dnf` are better aligned, right?

mwperina replied:
So packit use COPR to build RPMs, so `srpm_build_deps` are just packages, which needs to be installed to the plain system to be able to build SRPM. Once we have SRPM, COPR itself will install build dependencies as described in the spec file.

But integration tests are running on a testing farm and packit just uses testing farm API to run tests, but testing farm is independent to packit. Officially all packages required to run tests within tmt framework should be installed as a part of `prepare` phase of each plan using `packages` keyword: https://tmt.readthedocs.io/en/stable/spec/plans.html#prepare

Unfortunately tmt doesn't recognize properly ubuntu, so using `packages` keyword will break current GH CI, because `dnf install` doesn't work on ubuntu, that's why we install those packages in https://github.com/eclipse-bluechi/bluechi/blob/main/.github/workflows/integration-tests.yml#L49
For now I've used dnf only, because it's much easier that pip, but if we really need newer versions not in Fedora we could switch

engelmi replied:
Ah ok, thanks for the explanation! The packit + testing integration is quite complex

engelmi commented on tests/scripts/tests-setup.sh:
```diff
@@ -3,12 +3,32 @@
 fi
 
-podman build $ARCH -f ./containers/$CONTAINER_USED -t $BLUECHI_IMAGE_NAME .
+if [ "$CONTAINER_USED" = "integration-test-snapshot" ]; then
```
Not really part of this PR, but is this check really needed? Can't we just always detect `arch`?

mwperina replied:
Hmm, good catch, it's just a remain, which is no longer needed after https://github.com/eclipse-bluechi/bluechi/pull/441
I will remove it

mwperina committed:
Remove unneeded --arch parameter in test container builds

Passing `--arch` parameter to `podman build` in
`tests/scripts/tests-setup.sh` is no longer needed after
https://github.com/eclipse-bluechi/bluechi/pull/441 being merged.

Signed-off-by: Martin Perina <mperina@redhat.com>

engelmi reviewed (approved):
LGTM
